### DISCUSSION TITLE
Ideas for developing the roadmap for the project

### DISCUSSION THREAD

**Discussion by jarmos-san** [developersIndia/developersIndia#69] [Ideas]
### Feature details?

There's a template to work with, there are ideas & there are interested contributors willing to provide a helping hand to the project. What's missing is a "solid & proper" public roadmap to help the community work towards a common goal.

So, if you've ideas and/or experience on developing a roadmap for web-based applications, please feel free to discuss it here. You can share ideas, resources and/or suggestions below in the comment thread.

cc @siddhantk232

**Comment by jarmos-san** [comment]
I don't have much experience starting a web app project from project but here's pale attempt at developing a roadmap for this current project.

The gist is, if recreating [the Figma template](https://www.figma.com/file/jxpIq8GzilqzgyLpFBySP2/Developers-India-%3C3?node-id=5%3A58734) is the end-goal as-in `v1.x` of the project, then we should divide specific aspects of that end-goal into smaller achievable goals. For e.g. since the whole website consists of multiple section like the hero section, followed by the info section & then the goals sections & finally towards the end with a footer section. As such, it might be possible to development of individual or multiple sections of the website into multiple individuals roadmaps.

For further context, say developing the Hero section of the website would be in the roadmap for `v0.1.x`. The Info & Goals section into the roadmap for `v0.2.x` & so on. Obviously, this a very crude idea & the deciding what'll be included in to individual roadmaps will depend on the situation.

Regardless, I look forward to receiving feedback so that this idea can be improved further.

**Reply by afkcodes** [reply]
we will also be needing a mobile design i guess.

**Reply by jarmos-san** [reply]
> we will also be needing a mobile design i guess.

Good idea! We forgot about it dang it, but anyway I think we can add it to our `ROADMAP.md` as well & invite contributions. But if you can develop the template for mobile devices please let us know?

**Comment by siddhantk232** [comment]
A note on versioning for a web project like this.

I too was thinking smth very similar to [this answer](https://softwareengineering.stackexchange.com/a/366739) and I think we should not use semantic versioning (or at least ignore it for now) as we are not gonna be publishing to npmjs from developersindia/website.

It would make more sense to have smth like `v1` or `2022q1` so, you know, we can introduce new designs or enhancements that are bound to some special event (Christmas maybe).

Not to mention how easy it would make answering all those questions like "How should we bump the version this time now that we have a brand integration or a hackathon promotion."

**Reply by BlackGoku36** [reply]
Tbh, I don't really see a point in versioning a website, even if we have to, then +1 on versioning based on date. I suggest making it based on month like `2022.1` (January 2022) or maybe based on festivals/holidays like `Diwali 2022`. At the end, you wouldn't have to worry about when to bump version and to what version.

**Reply by jarmos-san** [reply]
+1 to that idea! I like it too & I bet the others will like as well!

@BlackGoku36 yup you're right semantic versioning doesn't really apply well to websites because a website is every changing. And no one expects it to have extreme backwards compatibility either.

That said, why don't you guys work together putting together a calendar-based roadmap then? I'm damn sure the rest of the community will be equally excited to work on a roadmap which is based on Indian festivities! Let me know if you guys additional support? If you send me a +1 then I'll convert this discussion into an issue thread & from there on you guys can collaborate developing the roadmap.

And above all, I'm excited to see the end results, so keep up the good work! :clinking_glasses:

**Reply by nirmalhk7** [reply]
Aside from actually building the website, what kind of content we expecting? Is there a roadmap for that part?

I'm guessing blogs would be a great starter point.

**Reply by jarmos-san** [reply]
@nirmalhk7 yes thanks to quite a lot of people who contributed to the discussion over here & on our Discord channels, I've had some ideas for developing a roadmap as well. I jotted them down elsewhere for safekeeping but its just that I'm not finding enough free to time to share a detailed write-up about it. Hopefully by the end of this week or on Sunday, I'll be able to draft the `ROADMAP.md` properly. So, you might want to subscribe to this repository for notifications!

About contents, we've nothing planned of that sort at least for this current year. We're busy maintaining the community as a whole let alone a community project, so we got our hands full. Unless someone from the community reaches out to us with a genuine interest in sharing community blog posts every now & then, I doubt the blog feature of the website will any light of the day.

**Reply by nirmalhk7** [reply]
I'd suggest you to take a look at similar projects by other communities online (I can give [IndiaInvestments](https://github.com/indiainvestments) as an example). 

I was thinking if not blogs, we can create a voting bot that posts well crafted responses (on any platform of this community) to the website. They'd be more like snippets but not blogs, but they'd still be very helpful.

**Reply by jarmos-san** [reply]
@Bhupesh-V take a look at the GitHub org @nirmalhk7 shared, it looks nice & we can definitely take some inspiration from them. As mentioned earlier in one of our discussions, we can make use of starboard entries (but it needs to be cleaned up & "refactored" first). As for what an ideal starboard should be like take a look at AnonCorp, it looks more professional & presentable to the general public.

**Reply by jarmos-san** [reply]
Also the secret gist you maintain could you note it down there? It sounds like a nice idea & we can definitely work towards it at some point in time.

**Comment by afkcodes** [comment]
let's break it down and make it something special to India, we can internally create tags based on semantic versioning.

**Reply by jarmos-san** [reply]
I'm sceptic about adding Semantic Versioning to a website project to be honest now that I read about @siddhantk232's suggestions. I think we should help him & @BlackGoku36 polish their idea further & help prepare a PR with a detailed `ROADMAP.md`.

P.S. I kinda like the idea about releasing based on calendar events like `Diwali 2020 Dhamaka` & so on! :laughing:

**Comment by jarmos-san** [comment]
### After Thoughts

Thanks to everyone who contributed to this discussion, I now have a good idea of what to include in the `ROADMAP.md`.

I gave it some thought & realised the "design" of the project's roadmap has been under our noses all this time! Like any other FOSS projects hosted on GitHub, we should've utilised "[GitHub milestones](https://docs.github.com/en/issues/using-labels-and-milestones-to-track-work/about-milestones)" right from the start! As such, I've perused through other big & established projects like [`winget`](https://github.com/microsoft/winget-cli) & how does it's maintainer(s) develop a roadmap as well.

### Why a Generic `ROADMAP.md` Won't Work For This Project

On that note, I decided to keep the roadmap & it's milestones simple enough for a couple of reasons:
1. [SemVer](https://semver.org/) doesn't apply to our project since backwards compatibility isn't much of a concern & above all, it's a simple website project!
2. Since SemVer doesn't apply to the project, it means we're on our own when it comes to building a roadmap. We're also aware of introducing unintentional complexities when starting from scratch & we don't want that. There's already so much stuff we've to handle, maintain & take care of.
3. We like & follow the "[_Simple is better than complex_](https://zen-of-python.info/simple-is-better-than-complex.html)" quote from the [Zen of Python](https://zen-of-python.info) on every aspects of our development practices.

### Proposal of What Our Roadmap's Milestones

On that note, I propose categorising each tasks into three broad milestones in our roadmap. And following are the milestones justifying why they're named as such:

- **Current** - for tracking current feature additions, events. projects & such which as been accepted by the community & has to be worked on immediately.
- **Next** - for tracking enhancements, projects, etc which are approved but the implementation detail is still under review & discussion.
- **Future** - for tracking feature and/or ideas which yet to be approved & are still receiving community feedback for polishing & further enhancement.

### About the Roadmap

Public roadmaps are meant to showcase a more general overview of a project's future development. As such, since no one can truly predict what could happen in the near future, there's no way to keep our roadmap static either. Hence, expect the roadmap to be ever-changing with time & the community as well.

The three milestones described above will be part of the overall roadmap & each of those milestones will contain smaller tasks to be completed (thanks @afkcodes for suggesting the idea! :heart:). For more context, if developing the landing page is part of the "_Current_" milestone, then we could include smaller individual tasks like "_developing the header section_" into a single issue thread & so on.

Issue threads which discusses a feature request and/or an enhancement which has been approved & received enough community feedback will be added to the _"Next"_ milestone. Similarly, the "Future" milestone will keep a track of backlog issues which are yet to be approved & has enough community feedback.

Till now, we've been only discussing how to track progress in the long-run, short-run progress should be tracked through [GitHub Releases](https://docs.github.com/en/repositories/releasing-projects-on-github/about-releases). Now we've already mentioned earlier SemVer isn't applicable to this current project as generic tags like this `vX.Y.Z` where `X`, `Y` & `Z` are positive integers, wouldn't work for our use case. To work around such a shortcoming @siddhantk232 & @BlackGoku36 suggested following a "_calendar release schedule_" ([see comment](https://github.com/developersIndia/website/discussions/69#discussioncomment-2130963)) instead.

But developersIndia doesn't have the dedicated manpower to work on strict calendar release schedule like [how GitPod.io does it](https://github.com/orgs/gitpod-io/projects/27). So, I feel releasing a tagged commit based on the year & month makes more sense. For more context, assuming the last commit which adds a functional feature to the landing page on Feb 2022, then we could release a version tagged as `v2022.2` & so on.

### Ending Notes

Everything about this roadmap is subject to change & as such please feel free to share your feedback on this PR #92. If things doesn't make sense right now or if it sounds too complicated, do not worry, everything will fall into place once the PR linked above is merged. And you can trust me, having a detailed roadmap in place will also help interested contributors collaborate on tasks more easily as well!

cc: @Bhupesh-V @yashsharan0805 @varshney-himanshu let me know if you guys have any feedback.

**Reply by nirmalhk7** [reply]
> Current - for tracking current feature additions, events. **projects** & such which as been accepted by the community & has to be worked on immediately.
Next - for tracking enhancements, **projects**, etc which are approved but the implementation detail is still under review & discussion.

Could you clarify what's 'project' here? Is it this project (i.e. the website) or is it any project this community undertakes on GitHub?

**Reply by jarmos-san** [reply]
> Could you clarify what's 'project' here? Is it this project (i.e. the website) or is it any project this community undertakes on GitHub?

Good catch, I generalised "project" over here to portray any individual functional ideas people might have to add to this project (the website). I should've rephrased it to something better but I guess I ran out of words. :laughing:

**Comment by neeravnadla** [comment]
Just one question. what are you actually trying to build here? team or community or brand?

**Reply by jarmos-san** [reply]
Hi @neeravnadla we're building a community & you're free to join us as well. Also if you've any questions you would like to ask us about the community please free to do so. We're on Reddit & Discord (for now) so do check out our GitHub org for more contact info.

**Reply by neeravnadla** [reply]
cool. So who's leading this project right now? or it's community-driven.

**Reply by jarmos-san** [reply]
It's community-driven but I volunteered to maintain & lead it (or at least till I find someone capable enough to take the responsibility on themselves).

**Reply by neeravnadla** [reply]
We better divide people's in teams. Otherwise it's gonna be Chaos managing people.

**Reply by jarmos-san** [reply]
That's not how community-maintained FOSS projects on GitHub works other than that yes, I'm aware of it & yes, I'm working on a solution as well. But generally speaking, contributors on GitHub never stick through the lifetime of the project which is why "_dividing people into teams_" won't work out.

Also, it appears this discussion is a bit off-topic so I suggest we continue talking about managing the project on Discord instead.

**Comment by jarmos-san** [comment]
Since we already have a solid ROADMAP in plan, we're closing this thread. If you've any further queries please reach out the moderators on GitHub.
